pkg/azurecluster/update.go

```diff
@@ -0,0 +1,60 @@
+package azurecluster
+
+import (
+	"context"
+
+	"github.com/giantswarm/k8sclient/v4/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/api/admission/v1beta1"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+
+	"github.com/giantswarm/azure-admission-controller/internal/errors"
+	"github.com/giantswarm/azure-admission-controller/internal/releaseversion"
+	"github.com/giantswarm/azure-admission-controller/pkg/validator"
+)
+
+type UpdateValidator struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+type UpdateValidatorConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+func NewUpdateValidator(config UpdateValidatorConfig) (*UpdateValidator, error) {
+	v := &UpdateValidator{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return v, nil
+}
+
+func (a *UpdateValidator) Validate(ctx context.Context, request *v1beta1.AdmissionRequest) (bool, error) {
+	azureClusterNewCR := &capzv1alpha3.AzureCluster{}
+	azureClusterOldCR := &capzv1alpha3.AzureCluster{}
+	if _, _, err := validator.Deserializer.Decode(request.Object.Raw, nil, azureClusterNewCR); err != nil {
+		return false, microerror.Maskf(errors.ParsingFailedError, "unable to parse AzureCluster CR: %v", err)
+	}
+	if _, _, err := validator.Deserializer.Decode(request.OldObject.Raw, nil, azureClusterOldCR); err != nil {
+		return false, microerror.Maskf(errors.ParsingFailedError, "unable to parse AzureCluster CR: %v", err)
+	}
+
+	oldClusterVersion, err := releaseversion.GetVersionFromCRLabels(azureClusterOldCR.Labels)
```

Comment:
We are calling `GetVersionFromCRLabels` (twice) and `UpgradeAllowed` from the `releaseversion` package. I was thinking if it would be better to add a new function `releaseversion.Validate()` and just call that? All the validations related to releases/versioning would be done there. If we were to add validations based on other things they would be in other packages. And the `UpdateValidator` would look something like

```Validate() {
    err = releaseversion.Validate()
    if err != nil {
        return false, microerror.mask(err)
    }

    err = anotherpackage.Validate()
    if err != nil {
        return false, microerror.mask(err)
    }
}```

Reply:
ok that makes sense

Reply:
I extracted the parsing of release to another shared module and renamed the validating function to Validate.
I don't like the idea of passing the CRs to the Validate function because it would mean it should know how to extract the release information from all CR types.
WDYT?

Reply:
Yeah, I see what you mean. `AzureConfig` and `AzureClusterConfig` are not labeled with the version right? If they are not, it's fine like you have it now.